Fabulous.XamarinForms/src/Fabulous.XamarinForms.Core/ViewUpdaters.fs

```diff
@@ -28,6 +28,120 @@ module ViewUpdaters =
+        | ValueSome prevColl, ValueSome newColl when identical prevColl newColl -> ()
+        | ValueSome prevColl, ValueSome newColl when prevColl <> null && newColl <> null && prevColl.Length = 0 && newColl.Length = 0 -> ()
+        | _, ValueNone -> clear ()
+        | _, ValueSome coll when (coll = null || coll.Length = 0) -> clear ()
+        | _, ValueSome coll ->
+            
+            let prevColl =
+                match prevCollOpt with
+                | ValueSome x -> x
+                | _ -> [||]
+
+            // Count the existing targetColl
+            let n = prevColl.Length
+            
+            let availableKeyedElements = Dictionary<string,'T>()
+            let rest= ResizeArray<'T>()
+            let newKeys = HashSet<string>()
+            
+            for i in 0 .. coll.Length-1 do
+                 let newChild = coll.[i]
+                 let key = getKey newChild
+                 match key with
+                 | ValueSome key -> newKeys.Add key |> ignore
+                 | ValueNone _ -> ()
+                 
+            for i in 0..n-1 do
+                let key = getKey prevColl.[i]
+                match key with
+                | ValueSome key ->
+                    availableKeyedElements.[key] <- prevColl.[i]
+                | ValueNone -> rest.Add prevColl.[i]
+            
+            let contains (key:string voption) =
+                match key with
+                | ValueSome key -> availableKeyedElements.ContainsKey key
+                | ValueNone -> false
+                
+            let newKeyContains el =
+                match getKey el with
+                | ValueSome key -> newKeys.Contains key
+                | ValueNone -> false
+  
+            // Adjust the existing targetColl and create the new targetColl
+            for i in 0 .. coll.Length-1 do
+                let newChild = coll.[i]
+                let key = getKey newChild
+                let elems =
+                        seq{
+                            yield! rest
+                            yield! availableKeyedElements.Values
+                           }
+                if (prevColl|>Seq.exists (identical newChild)) then
+                    move i newChild
+                    match getKey newChild with
+                     | ValueSome key -> availableKeyedElements.Remove key |> ignore
+                     |_ ->rest.Remove newChild |> ignore
+                     
+                elif (key.IsSome)
+                then
+                  if(contains (key)) then 
+                     let previousKeyedElement = availableKeyedElements.[key.Value] // We should be safe here
+                     if(canReuse previousKeyedElement newChild) then
+                      update i previousKeyedElement newChild
+                      availableKeyedElements.Remove key.Value |> ignore
+                  else
+                     match (elems|>Seq.tryFind(fun c-> canReuse c newChild)) with
```

Comment:
You could directly search `rest` here instead of `elems`.
With the 2nd comment I made, `availableKeyedElements` will only contain elements reused inside `if contains key`.

Reply:
OK